### PR TITLE
Add AWS AutoScaling lifecycle hooks

### PR DESCRIPTION
**Pull request by mwest1066**
- [x] Get the AWS region from the instance metadata, rather than hard-coding it.
- [x] Fix the fetching of instance-id from metadata.
- [x] Store instance config in SecretsManager, with a name given by a tag on the instance. This avoids building config information into the AMI and means we can use the same AMI for staging and prod. We have two different configs stored in SecretsManager, and we start the instance with the appropriate tag so it knows which one to use.
- [x] Add “starting” lifecycle hook.
- [ ] Add “terminating” lifecycle hook.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -15,10 +16,11 @@ const jobLogger = require('./lib/jobLogger');
 const configManager = require('./lib/config');
 const config = require('./lib/config').config;
 const healthCheck = require('./lib/healthCheck');
+const lifecycle = require('./lib/lifecycle');
 const pullImages = require('./lib/pullImages');
 const receiveFromQueue = require('./lib/receiveFromQueue');
 const timeReporter = require('./lib/timeReporter');
-const util = require('./lib/util');
+const dockerUtil = require('./lib/docker-util');
```
Comment (nwalters512) [inline index.js]: Nit: would prefer `./lib/dockerUtil` to match the casing of our other util modules.

File: index.js [hunk 2]
```diff
@@ -52,7 +57,7 @@ async.series([
     },
     (callback) => {
         if (!config.useDatabase || !config.reportLoad) return callback(null);
-        load.init(config.maxConcurrentJobs);
+        load.init();
```
Comment (nwalters512) [inline index.js]: Could you elaborate a bit on why this change was made?

File: index.js [hunk 3]
```diff
@@ -84,16 +97,20 @@ async.series([
                       success();
                   });
               }, (err) => {
-                  if (ERR(err, (err) => globalLogger.error(err)));
+                  if (ERR(err, (err) => globalLogger.error('receive error:', err)));
```
Comment (nwalters512) [inline index.js]: Have you tested that our logger will gracefully handle multiple arguments? I don't necessarily trust Winston anymore 🙃

File: lib/config.js [hunk 4]
```diff
@@ -33,29 +34,93 @@ config.loadConfig = function(callback) {
...
                     }
+                    callback(null);
+                }
+            });
+        },
+        (callback) => {
+            if (!exportedConfig.runningInEc2) return callback(null);
+
+            // if we are inside EC2, look up a special tag and use its value to
+            // find a secret containing configuration data
```
Comment (nwalters512) [inline lib/config.js]: Could you elaborate a bit in the comments what the shape of this secret data is?

File: index.js (hunk 2 above)
Comment (mwest1066) [reply index.js]: The intent here is to improve our CloudWatch output, so that `maxJobs` for an entire grader fleet will show the maximum number of jobs that we are able to handle at any given time. That is, the number of `currentJobs` for the entire fleet should range between 0 and `maxJobs`, and should be equal to `maxJobs` when all graders are 100% busy.

When a grader starts up and before it pulls images, it isn’t able to process jobs and so we don’t want to include it’s `maxConcurrentJobs` in the calculation of `maxJobs`.

Now that I write this out, it occurs to me that the other way we could do this is to filter on `lifecycle_state = ‘InService’` when summing `maxJobs` for CloudWatch. I made this early change before I added the recording of `lifecycle_state`.

What do you think? Would it be cleaner to return this to always report the `maxConcurrentJobs` that we are going to be able to handle once we are `InService`, and filter in the CloudWatch reporter?

File: index.js (hunk 3 above)
Comment (mwest1066) [reply index.js]: This is actually all because I believe Winston (which I still hate) needs to have a string as the first argument. If you give it an object as the first argument, it casts it to a string, which is what results in the `[object Object]` logs. If you give it `(string, object)` then it prints the string and `JSON.stringifies` the object. At least it approximately does this. It also looks for `object.message` and appends that to the string, and executes various other irritating functionality. I really hate Winston! Why can’t it just print the errors?! 😄 

Anyway, based on this reading of the Winston docs, I‘ve been going through and adding strings to as the first argument of all logger calls. I should note that I’m only 80% confident about what Winston is actually doing, unfortunately, so I’m happy to investigate further.

File: lib/config.js (hunk 4 above)
Comment (mwest1066) [reply lib/config.js]: Good point. It’s JSON in the same format as our `config` objects. I’ll leave this conversation here open to remind me to fix the comment.

File: index.js (hunk 1 above)
Comment (mwest1066) [reply index.js]: done

File: index.js (hunk 2 above)
Comment (mwest1066) [reply index.js]: I reverted this change and implemented it by filtering in the PL stats sproc. I think this is better because we are reporting more information (how many jobs will we be able to run, and are we healthy/in-service) and letting the final report aggregator figure out what they want to compute.

File: lib/config.js (hunk 4 above)
Comment (mwest1066) [reply lib/config.js]: done

File: index.js (hunk 3 above)
Comment (mwest1066) [reply index.js]: I confirmed. If you call winston like `logger.info({a:1, b:2})` then you just get `[object Object]`. It needs the first argument to be a string, to force it to format the second argument as JSON. Unfortunately the two-argument form _requires_ the second argument to be a regular object to display it in a useful way. If you call `logger.info(‘first’, ‘second’)` then only `first` shows up in the log. Argh!!!

In any case, for now I’m adding the first string argument to (hopefully) get meaningful output.

**Comment by mwest1066** [comment]
I’m going to merge/deploy this without the “terminating” lifecycle hook. This is more complex (needs an SNS endpoint and listeners) and is less important for stability than the “launching” hook.
